### ISSUE TITLE
fakeAsync test fails when testing code using rxjs 5 rxjs/scheduler/asap

### ISSUE THREAD

**Issue by asgallant** [angular/zone.js#1061]
`fakeAsync` tests throw the error "Unknown macroTask scheduled in fake async test: setImmediate" when the code under test uses the [rxjs Asap Scheduler](http://reactivex.io/rxjs/variable/index.html#static-variable-asap).

Steps to reproduce:

1. Write function using Asap Scheduler

```
import { asap } from 'rxjs/scheduler/asap';

export class TestClass {
  complete = false;
  doSomethingAsap() {
    asap.schedule(() => {
      this.complete = true;
    });
  }
}
```

2. Write unit test against the function

```
import { fakeAsync, tick } from '@angular/core/testing';
import { TestClass } from './test-class';

describe('TestClass', () => {
  it('should do stuff asynchronously', fakeAsync(() => {
    // arrange
    const test = new TestClass();

    // act
    test.doSomethingAsap();

    tick();

    // assert
    expect(test.complete).toBeTruthy();
  }));
});
```

Expected results: test passes
Actual results: test throws "Unknown macroTask scheduled in fake async test: setImmediate"

Found using Angular 5.2.9 and Zone.js 0.8.20

**Comment by JiaLiPassion** [comment]
@asgallant , thank you for posting the issue, I will fix this one.

**Comment by tao-cumplido** [comment]
@JiaLiPassion 
Is this supposed to work in 0.8.26 already?
I just updated but I am still experiencing the issue, or do I need to update Angular as well?

I have a test that looks kinda like this
```typescript
it('...', fakeAsync(() => {
    // mock server exposes subject, triggers nested callbacks with setImmediate
    server.subject.next('foo');
    tick();
    expect(itemToTest).toMatch('foo');
}));
```
The above test fails, while the one below succeeds
```typescript
it('...', async(() => {
    // mock server exposes subject, triggers nested callbacks with setImmediate
    server.subject.next('foo');
    setImmediate(() => expect(itemToTest).toMatch('foo'));
}));
```

**Comment by JiaLiPassion** [comment]
@tao-cumplido, could you provide a reproduce repo?

**Comment by tao-cumplido** [comment]
@JiaLiPassion 
Sure, I can try. It will probably take me a few days though. I will post it here as soon as I have it ready.

**Comment by tao-cumplido** [comment]
@JiaLiPassion 
I haven't had time yet to try to isolate the problem, but I just tried to write a router redirection test with `fakeAsync` and again it didn't work where regular async/await did. So, I am not sure if maybe I got something fundamentally wrong in using `fakeAsync` and thought I'd ask before trying to further investigate the example above.

So, I wrote the following test and the version with `fakeAsync` fails while the one without succeeds.

```typescript
it(`should redirect to '/help' for invalid routes`, fakeAsync(() => {
    router.navigate(['foo']);
    tick();
    expect(location.path()).toEqual('/help');
}));
```

```typescript
it(`should redirect to '/help' for invalid routes`, async () => {
    await router.navigate(['foo']);
    expect(location.path()).toEqual('/help');
});
```

Should they both do the same thing, or am I using `fakeAsync` wrong?
I got the `fakeAsync` example from [here](https://codecraft.tv/courses/angular/unit-testing/routing/#_testing_routing).
